subprojects/gdk_rust/gdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,1212 @@
+                self.data_root = net_params["state_dir"]
+                    .as_str()
+                    .map(|x| x.to_string())
+                    .unwrap_or_else(|| "".into());
+                info!("setting db_root to {:?}", self.data_root);
+            }
+
+            let mnemonic = match self.get_mnemonic() {
+                Ok(mnemonic) => Some(mnemonic.clone()),
+                Err(_) => None,
+            };
+            match mnemonic {
+                Some(mnemonic) => self.login(&mnemonic, None).map(|_| ())?,
+                None => self.state = State::Connected,
+            }
+        }
+        Ok(())
+    }
+
+    fn disconnect(&mut self) -> Result<(), Error> {
+        info!("disconnect state:{:?}", self.state);
+        if self.state != State::Disconnected {
+            self.closer.close()?;
+            self.state = State::Disconnected;
+        }
+        Ok(())
+    }
+
+    fn login(
+        &mut self,
+        mnemonic: &Mnemonic,
+        password: Option<Password>,
+    ) -> Result<Vec<Notification>, Error> {
+        info!("login {:?} {:?}", self.network, self.state);
+        if self.state == State::Logged {
+            return Ok(vec![]);
+        }
+
+        // TODO: passphrase?
+
+        let mnem_str = mnemonic.clone().get_mnemonic_str();
+        let seed = wally::bip39_mnemonic_to_seed(
+            &mnem_str,
+            &password.map(|p| p.get_password_str()).unwrap_or_default(),
+        )
+        .ok_or(Error::InvalidMnemonic)?;
+        let secp = Secp256k1::new();
+        let xprv =
+            ExtendedPrivKey::new_master(bitcoin::network::constants::Network::Testnet, &seed)?;
+
+        // BIP44: m / purpose' / coin_type' / account' / change / address_index
+        // coin_type = 0 bitcoin, 1 testnet, 1776 liquid bitcoin as defined in https://github.com/satoshilabs/slips/blob/master/slip-0044.md
+        // slip44 suggest 1 for every testnet, so we are using it also for regtest
+        let coin_type: u32 = match self.network.id() {
+            NetworkId::Bitcoin(bitcoin_network) => match bitcoin_network {
+                bitcoin::Network::Bitcoin => 0,
+                bitcoin::Network::Testnet => 1,
+                bitcoin::Network::Regtest => 1,
+            },
+            NetworkId::Elements(elements_network) => match elements_network {
+                ElementsNetwork::Liquid => 1776,
+                ElementsNetwork::ElementsRegtest => 1,
+            },
+        };
+        // since we use P2WPKH-nested-in-P2SH it is 49 https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki
+        let path_string = format!("m/49'/{}'/0'", coin_type);
+        info!("Using derivation path {}/0|1/*", path_string);
+        let path = DerivationPath::from_str(&path_string)?;
+        let xprv = xprv.derive_priv(&secp, &path)?;
+        let xpub = ExtendedPubKey::from_private(&secp, &xprv);
+
+        let wallet_desc = format!("{}{:?}{}", xpub, self.network, DB_VERSION);
```

Comment:
Hashing the xpub in the filename allows an adversary to make `is the wallet on the device using the xpub <XYZ>?` tests against encrypted wallets.

If a user unintentionally leaks his xpub and an adversary suspects that he's related to it, this could be used to definitively tie him to the wallet's activity despite the wallet being locked.

Reply:
yes, it's not perfect, do you have any alternative? (we want to support watch-only wallet)

Reply:
Can you explain me a bit more about how the xpub is used to enable this?